fast64_internal/f3d/f3d_material.py

```diff
@@ -177,85 +177,144 @@ def update_draw_layer(self, context):
         set_output_node_groups(material)
 
 
-def rendermodePresetToBits(rdp_settings: "RDPSettings") -> Tuple[int, int]:
-    f3d = get_F3D_GBI()
-    r1 = getattr(f3d, rdp_settings.rendermode_preset_cycle_1, f3d.G_RM_AA_ZB_OPA_SURF)
-    r2 = getattr(f3d, rdp_settings.rendermode_preset_cycle_2, f3d.G_RM_AA_ZB_OPA_SURF)
-    if rdp_settings.g_mdsft_cycletype == "G_CYC_1CYCLE":
-        r2 = 0
-    return r1, r2
-
-
-def all_blender_uses(rdp_settings: "RDPSettings") -> Dict[str, bool]:
+def rendermode_preset_to_advanced(material: bpy.types.Material):
     """
-    Returns a dictionary of the external features which the blender may or may
-    not use, or None if set_rendermode is disabled so we don't know.
+    Set all individual controls for the rendermode from the preset rendermode.
     """
-    if not rdp_settings.set_rendermode:
-        return None
-    is_one_cycle = rdp_settings.g_mdsft_cycletype == "G_CYC_1CYCLE"
-    if rdp_settings.rendermode_advanced_enabled:
-        useZ = rdp_settings.z_cmp or rdp_settings.z_upd
-        useShade = rdp_settings.blend_a1 == "G_BL_A_SHADE"
-        if not is_one_cycle:
-            useShade = useShade or rdp_settings.blend_a2 == "G_BL_A_SHADE"
+    settings = material.f3d_mat.rdp_settings
+    f3d = get_F3D_GBI()
+    
+    if settings.rendermode_advanced_enabled:
+        # Already in advanced mode, don't overwrite this with the preset
+        return
+    
+    def get_with_default(preset, default):
+        # Use the default either if we are not setting rendermode, or if the
+        # preset is not in the GBI. We do want to set the advanced settings
+        # even if not setting rendermode, because they are read for nodes preview.
+        if not settings.set_rendermode:
+            return default
+        return getattr(f3d, preset, default)
+    
+    is_one_cycle = settings.g_mdsft_cycletype == "G_CYC_1CYCLE"
+    if is_one_cycle:
+        r = get_with_default(settings.rendermode_preset_cycle_1, f3d.G_RM_AA_ZB_OPA_SURF)
+        r1 = r
+        # For GBL_c2, look at the same bits as for cycle 1. Cycle 1 bits are copied
+        # to cycle 2 bits at export.
+        r2 = r >> 2
```

Comment:
I feel like `r2 = r >> 2` is an unclear hack, and I've been writing more text explaining it than it takes to just make the set-blender-coefficients different based on cycle type

I suggest

```
        # In 1-cycle mode both cycles are the same
        r1 = r
        settings.blend_p1 = f3d.blendColorDict[(r1 >> 30) & 3]
        settings.blend_a1 = f3d.blendAlphaDict[(r1 >> 26) & 3]
        settings.blend_m1 = f3d.blendColorDict[(r1 >> 22) & 3]
        settings.blend_b1 = f3d.blendMixDict[(r1 >> 18) & 3]
        settings.blend_p2 = settings.blend_p1
        settings.blend_a2 = settings.blend_a1
        settings.blend_m2 = settings.blend_m1
        settings.blend_b2 = settings.blend_b1
```

and move the code below for setting pamb12 to inside the 2-cycle else

Reply:
Gonna have to disagree on this one--I strongly prefer one line of code which ends up being one instruction to 8 lines of code where four of them are repeats of the above. I've improved the comment though.